File: lib-integration-tests/src/test/java/io/github/kory33/guardedqueries/core/testharnesses/RewriterEquivalenceTestHarness.java
package io.github.kory33.guardedqueries.core.testharnesses;

import io.github.kory33.guardedqueries.core.datalog.DatalogRewriteResult;
import io.github.kory33.guardedqueries.core.formalinstance.FormalInstance;
import io.github.kory33.guardedqueries.core.rewriting.GuardedRuleAndQueryRewriter;
import io.github.kory33.guardedqueries.core.subsumption.formula.MinimalExactBodyDatalogRuleSet;
import io.github.kory33.guardedqueries.core.subsumption.formula.MinimallyUnifiedDatalogRuleSet;
import io.github.kory33.guardedqueries.core.testcases.GTGDRuleAndConjunctiveQuery;
import io.github.kory33.guardedqueries.core.testcases.GTGDRuleAndGTGDReducibleQuery;
import uk.ac.ox.cs.pdq.fol.Atom;
import uk.ac.ox.cs.pdq.fol.Constant;
import uk.ac.ox.cs.pdq.fol.Predicate;
import uk.ac.ox.cs.pdq.fol.Variable;

import java.time.Instant;
import java.util.Date;

import static io.github.kory33.guardedqueries.core.testharnesses.InstanceGeneration.randomInstanceOver;

public record RewriterEquivalenceTestHarness(
        GuardedRuleAndQueryRewriter firstRewriter,
        GuardedRuleAndQueryRewriter secondRewriter
) {
    private static void logWithTime(String message) {
        System.out.println("[" + Date.from(Instant.now()) + "] " + message);
    }

    private record RewriteResultsToBeCompared(
            DatalogRewriteResult resultFromFirstRewriter,
            DatalogRewriteResult resultFromSecondRewriter,
            // the atom to which we can materialize computed homomorphisms
            Atom answerAtom
    ) {
        public FormalInstance<Constant> answerWithFirstRewriter(
                final FormalInstance<Constant> testInstance
        ) {
            return RunOutputDatalogProgram.answersOn(testInstance, resultFromFirstRewriter, answerAtom, c -> c);
        }

        public FormalInstance<Constant> answerWithSecondRewriter(
                final FormalInstance<Constant> testInstance
        ) {
            return RunOutputDatalogProgram.answersOn(testInstance, resultFromSecondRewriter, answerAtom, c -> c);
        }
    }

    private DatalogRewriteResult minimizeRewriteResultAndLogIntermediateCounts(
            DatalogRewriteResult originalRewriteResult
    ) {
        logWithTime("# of subgoal derivation rules in original output: " +
                originalRewriteResult.subgoalAndGoalDerivationRules().rules().size());

        final var minimalExactBodyMinimizedRewriting = originalRewriteResult
                .minimizeSubgoalDerivationRulesUsing(MinimalExactBodyDatalogRuleSet::new);

        logWithTime("# of subgoal derivation rules in minimalExactBodyMinimizedRewriting: " +
                minimalExactBodyMinimizedRewriting.subgoalAndGoalDerivationRules().rules().size());

        final var minimizedRewriting = minimalExactBodyMinimizedRewriting
                .minimizeSubgoalDerivationRulesUsing(MinimallyUnifiedDatalogRuleSet::new);

        logWithTime("# of subgoal derivation rules in minimizedRewriting: " +
                minimizedRewriting.subgoalAndGoalDerivationRules().rules().size());

        return minimizedRewriting;
    }

    private RewriteResultsToBeCompared rewriteUsingTwoRewriters(
            final GTGDRuleAndConjunctiveQuery ruleQuery
    ) {
        logWithTime("Rewriting " + ruleQuery.query());

        logWithTime("Rewriting with " + firstRewriter + "...");
        final var firstRewritingStart = System.nanoTime();
        final var firstRewriting = firstRewriter.rewrite(ruleQuery.guardedRules(), ruleQuery.query());
        logWithTime("Done rewriting with " + firstRewriter + " in " + (System.nanoTime() - firstRewritingStart) + "ns");
        final var minimizedFirstRewriting = minimizeRewriteResultAndLogIntermediateCounts(firstRewriting);

        logWithTime("Rewriting with " + secondRewriter + "...");
        final var secondRewritingStart = System.nanoTime();
        final var secondRewriting = secondRewriter.rewrite(ruleQuery.guardedRules(), ruleQuery.query());
        logWithTime("Done rewriting with " + secondRewriter + " in " + (System.nanoTime() - secondRewritingStart) + "ns");
        final var minimizedSecondRewriting = minimizeRewriteResultAndLogIntermediateCounts(secondRewriting);

        final var answerAtom = Atom.create(
                Predicate.create("Answer", ruleQuery.deduplicatedQueryFreeVariables().size()),
                ruleQuery.deduplicatedQueryFreeVariables().toArray(Variable[]::new)
        );

        return new RewriteResultsToBeCompared(minimizedFirstRewriting, minimizedSecondRewriting, answerAtom);
    }

    /**
     * Test that the two {@link GuardedRuleAndQueryRewriter}s produce
     * equivalent Datalog rewritings on the given {@link GTGDRuleAndGTGDReducibleQuery}.
     * <p>
     * The test is repeatedly performed on randomly generated database instances
     * (with the number of test rounds being specified by {@code instanceGenerationRoundCount}).
     */
    public void checkThatTwoRewritersAgreeOn(
            final GTGDRuleAndConjunctiveQuery ruleQuery,
            final int instanceGenerationRoundCount
    ) {
        final var rewritings = rewriteUsingTwoRewriters(ruleQuery);

        for (int i = 0; i < instanceGenerationRoundCount; i++) {
            final var testInstance = randomInstanceOver(ruleQuery.signature());

            final var firstAnswer = rewritings.answerWithFirstRewriter(testInstance);
            final var secondAnswer = rewritings.answerWithSecondRewriter(testInstance);

            if (!firstAnswer.equals(secondAnswer)) {
                throw new AssertionError("Two rewriters gave different answers! " +
                        "input = " + testInstance + ", " +
                        "first rewriter answer = " + firstAnswer + ", " +
                        "second rewriter answer = " + secondAnswer + ", " +
                        "first rewriter = " + firstRewriter + ", " +
                        "second rewriter = " + secondRewriter
                );
            } else {
                logWithTime("Test " + i + " passed, " +
                        "input size = " + testInstance.facts.size() + ", " +
                        "answer size = " + firstAnswer.facts.size()
                );
            }
        }
    }

    /**
     * Test that the two {@link GuardedRuleAndQueryRewriter}s produce
     * equivalent Datalog rewritings on the given {@link GTGDRuleAndGTGDReducibleQuery}.
     * <p>
     * The test is repeatedly performed on randomly generated database instances
     * (with the number of test rounds being specified by {@code instanceGenerationRoundCount}).
     */
    public void checkThatTwoRewritersAgreeOn(
            final GTGDRuleAndGTGDReducibleQuery ruleQuery,
            final int instanceGenerationRoundCount
    ) {
        checkThatTwoRewritersAgreeOn(ruleQuery.asGTGDRuleAndConjunctiveQuery(), instanceGenerationRoundCount);
    }
}

File: lib-integration-tests/src/test/scala/io/github/kory33/guardedqueries/core/testharnesses/RewriterEquivalenceTestHarness.scala
package io.github.kory33.guardedqueries.core.testharnesses

import io.github.kory33.guardedqueries.core.datalog.DatalogRewriteResult
import io.github.kory33.guardedqueries.core.formalinstance.FormalInstance
import io.github.kory33.guardedqueries.core.rewriting.GuardedRuleAndQueryRewriter
import io.github.kory33.guardedqueries.core.subsumption.formula.MinimalExactBodyDatalogRuleSet
import io.github.kory33.guardedqueries.core.subsumption.formula.MinimallyUnifiedDatalogRuleSet
import io.github.kory33.guardedqueries.core.testcases.GTGDRuleAndConjunctiveQuery
import io.github.kory33.guardedqueries.core.testcases.GTGDRuleAndGTGDReducibleQuery
import io.github.kory33.guardedqueries.core.testharnesses.InstanceGeneration.randomInstanceOver
import uk.ac.ox.cs.pdq.fol.Atom
import uk.ac.ox.cs.pdq.fol.Constant
import uk.ac.ox.cs.pdq.fol.Predicate
import uk.ac.ox.cs.pdq.fol.Term
import uk.ac.ox.cs.pdq.fol.Variable

import java.time.Instant
import java.util.Date
import scala.jdk.CollectionConverters._

object RewriterEquivalenceTestHarness {
  private def logWithTime(message: String): Unit = {
    System.out.println(s"[${Date.from(Instant.now)}] $message")
  }

  private case class RewriteResultsToBeCompared(
    resultFromFirstRewriter: DatalogRewriteResult,
    resultFromSecondRewriter: DatalogRewriteResult,
    answerAtom: Atom
  ) {
    def answerWithFirstRewriter(testInstance: FormalInstance[Constant])
      : FormalInstance[Constant] = RunOutputDatalogProgram.answersOn(
      testInstance,
      resultFromFirstRewriter,
      answerAtom,
      (c: Constant) => c
    )

    def answerWithSecondRewriter(testInstance: FormalInstance[Constant])
      : FormalInstance[Constant] = RunOutputDatalogProgram.answersOn(
      testInstance,
      resultFromSecondRewriter,
      answerAtom,
      (c: Constant) => c
    )
  }
}

case class RewriterEquivalenceTestHarness(firstRewriter: GuardedRuleAndQueryRewriter,
                                          secondRewriter: GuardedRuleAndQueryRewriter
) {
  private def minimizeRewriteResultAndLogIntermediateCounts(
    originalRewriteResult: DatalogRewriteResult
  ) = {
    RewriterEquivalenceTestHarness.logWithTime(
      "# of subgoal derivation rules in original output: " + originalRewriteResult.subgoalAndGoalDerivationRules.rules.size
    )

    val minimalExactBodyMinimizedRewriting =
      originalRewriteResult.minimizeSubgoalDerivationRulesUsing(() =>
        MinimalExactBodyDatalogRuleSet()
      )

    RewriterEquivalenceTestHarness.logWithTime(
      "# of subgoal derivation rules in minimalExactBodyMinimizedRewriting: " + minimalExactBodyMinimizedRewriting.subgoalAndGoalDerivationRules.rules.size
    )

    val minimizedRewriting =
      minimalExactBodyMinimizedRewriting.minimizeSubgoalDerivationRulesUsing(() =>
        MinimallyUnifiedDatalogRuleSet()
      )

    RewriterEquivalenceTestHarness.logWithTime(
      "# of subgoal derivation rules in minimizedRewriting: " + minimizedRewriting.subgoalAndGoalDerivationRules.rules.size
    )

    minimizedRewriting
  }

  private def rewriteUsingTwoRewriters(ruleQuery: GTGDRuleAndConjunctiveQuery) = {
    RewriterEquivalenceTestHarness.logWithTime(s"Rewriting ${ruleQuery.query}")
    RewriterEquivalenceTestHarness.logWithTime(s"Rewriting with $firstRewriter...")

    val firstRewritingStart = System.nanoTime
    val firstRewriting = firstRewriter.rewrite(ruleQuery.guardedRules, ruleQuery.query)
    RewriterEquivalenceTestHarness.logWithTime(
      s"Done rewriting with $firstRewriter in ${System.nanoTime - firstRewritingStart}ns"
    )

    val minimizedFirstRewriting = minimizeRewriteResultAndLogIntermediateCounts(firstRewriting)
    RewriterEquivalenceTestHarness.logWithTime(s"Rewriting with $secondRewriter...")
    val secondRewritingStart = System.nanoTime
    val secondRewriting = secondRewriter.rewrite(ruleQuery.guardedRules, ruleQuery.query)
    RewriterEquivalenceTestHarness.logWithTime(
      s"Done rewriting with $secondRewriter in ${System.nanoTime - secondRewritingStart}ns"
    )
    val minimizedSecondRewriting =
      minimizeRewriteResultAndLogIntermediateCounts(secondRewriting)

    val answerAtom = Atom.create(
      Predicate.create("Answer", ruleQuery.deduplicatedQueryFreeVariables.size),
      ruleQuery.deduplicatedQueryFreeVariables.toArray[Term]: _*
    )

    RewriterEquivalenceTestHarness.RewriteResultsToBeCompared(
      minimizedFirstRewriting,
      minimizedSecondRewriting,
      answerAtom
    )
  }

  /**
   * Test that the two [[GuardedRuleAndQueryRewriter]]s produce equivalent Datalog rewritings on
   * the given [[GTGDRuleAndConjunctiveQuery]].
   *
   * The test is repeatedly performed on randomly generated database instances (with the number
   * of test rounds being specified by <pre>instanceGenerationRoundCount</pre>).
   */
  def checkThatTwoRewritersAgreeOn(ruleQuery: GTGDRuleAndConjunctiveQuery,
                                   instanceGenerationRoundCount: Int
  ): Unit = {
    val rewritings = rewriteUsingTwoRewriters(ruleQuery)
    for (i <- 0 until instanceGenerationRoundCount) {
      val testInstance = randomInstanceOver(ruleQuery.signature)
      val firstAnswer = rewritings.answerWithFirstRewriter(testInstance)
      val secondAnswer = rewritings.answerWithSecondRewriter(testInstance)
      if (!(firstAnswer == secondAnswer)) throw new AssertionError(
        "Two rewriters gave different answers! " + "input = " + testInstance + ", " + "first rewriter answer = " + firstAnswer + ", " + "second rewriter answer = " + secondAnswer + ", " + "first rewriter = " + firstRewriter + ", " + "second rewriter = " + secondRewriter
      )
      else if (i % 20 == 0) RewriterEquivalenceTestHarness.logWithTime(
        "Test " + i + " passed, " + "input size = " + testInstance.facts.size + ", " + "answer size = " + firstAnswer.facts.size
      )
    }
  }

  /**
   * Test that the two [[GuardedRuleAndQueryRewriter]]s produce equivalent Datalog rewritings on
   * the given [[GTGDRuleAndGTGDReducibleQuery]].
   *
   * The test is repeatedly performed on randomly generated database instances (with the number
   * of test rounds being specified by <pre>instanceGenerationRoundCount</pre>).
   */
  def checkThatTwoRewritersAgreeOn(ruleQuery: GTGDRuleAndGTGDReducibleQuery,
                                   instanceGenerationRoundCount: Int
  ): Unit = {
    checkThatTwoRewritersAgreeOn(
      ruleQuery.asGTGDRuleAndConjunctiveQuery,
      instanceGenerationRoundCount
    )
  }
}
